src/natcap/invest/coastal_blue_carbon/coastal_blue_carbon.py

```diff
@@ -188,941 +187,1485 @@
+
+    taskgraph_cache_dir = os.path.join(
+        args['workspace_dir'], TASKGRAPH_CACHE_DIR_NAME)
+    task_graph = taskgraph.TaskGraph(
+        taskgraph_cache_dir, n_workers)
+
+    suffix = args.get('suffix', '')
+    intermediate_dir = os.path.join(
+        args['workspace_dir'], INTERMEDIATE_DIR_NAME)
+    output_dir = os.path.join(
+        args['workspace_dir'], OUTPUT_DIR_NAME)
+
+    transition_years = set([int(year) for year in args['transition_years']])
+    disturbance_magnitude_rasters = args['disturbance_magnitude_rasters']
+    half_life_rasters = args['half_life_rasters']
+    yearly_accum_rasters = args['annual_rate_of_accumulation_rasters']
+
+    prices = None
+    if 'carbon_prices_per_year' in args and args['carbon_prices_per_year']:
+        prices = {int(year): float(price)
+                  for (year, price) in args['carbon_prices_per_year'].items()}
+    discount_rate = float(args['discount_rate'])
+    baseline_lulc_year = int(args['baseline_lulc_year'])
+
+    # ASSUMPTIONS
+    #
+    # This rebuild assumes that this timeseries analysis is ONLY taking place
+    # for the transitions at hand.  Everything that happens between the
+    # baseline year and the first transition isn't all that interesting since
+    # the only thing that can happen is accumulation. That can be modeled with
+    # a few raster calculator operations.  Everything within this loop is way
+    # more interesting and tricky to get right, hence the need for an extra
+    # function.
+    stock_rasters = {
+        (min(transition_years) - 1): {
+            POOL_SOIL: args['stocks_at_first_transition'][POOL_SOIL],
+            POOL_BIOMASS: args['stocks_at_first_transition'][POOL_BIOMASS],
+            POOL_LITTER: args['stocks_at_first_transition'][POOL_LITTER],
+        }
+    }
+    net_sequestration_rasters = {
+        (min(transition_years) - 1): {
+            POOL_SOIL: args['annual_rate_of_accumulation_rasters'][POOL_SOIL],
+            POOL_BIOMASS: (
+                args['annual_rate_of_accumulation_rasters'][POOL_BIOMASS]),
+            POOL_LITTER: (
+                args['annual_rate_of_accumulation_rasters'][POOL_LITTER]),
+        }
+    }
+    disturbance_vol_rasters = {}
+    emissions_rasters = {}
+    year_of_disturbance_rasters = {}
+    total_carbon_rasters = {}
+    prior_transition_year = None
+    current_transition_year = None
+
+    current_disturbance_vol_tasks = {}
+    prior_stock_tasks = {}
+    current_year_of_disturbance_tasks = {}
+    current_emissions_tasks = {}
+    prior_net_sequestration_tasks = {}
+    current_net_sequestration_tasks = {}
+    valuation_tasks = {}
+
+    first_transition_year = min(transition_years)
+    final_year = int(args['analysis_year'])
+
+    summary_net_sequestration_tasks = []
+    summary_net_sequestration_raster_paths = {
+        first_transition_year: args['sequestration_since_baseline_raster']}
+
+    for year in range(first_transition_year, final_year+1):
```

Comment:
Curious why this loop is over every year in the range, rather than just the transition years? Maybe could avoid some extra intermediate rasters if this could be done only on transition years

Reply:
This was absolutely a consideration, and, in fact, the implementation that this replaces does almost exactly that.  Certainly the prior model only wrote rasters for each of the transitions, containing as much as possible in numpy arrays in memory.

There were 3 major issues that arose from this approach:

1. **memory constraints**
  This model has no upper bound on how many years the timeseries can operate over.  So even if we were to have each task operating on only those years between transitions, even that number could be very large and exhaust available memory.  We had many issues with memory usage in the prior model, and iterating over lots of files is the safest general-case solution for a timeseries model with an unbounded number of years.
2. **in practice, the intermediate rasters are useful for analyses**
  The prior implementation wrote as few rasters out as it could, which ultimately ended up making it harder for some of our partner organizations to to the science they needed to do.  By providing all these intermediate rasters, we allow our users to do their own analytics and summaries on how carbon stocks change over time, which would have been game-changing to our colleagues in analyses past.
3. **debugging is easier with rasters than with multidimensional arrays**
  The prior implementation used 3D arrays (`m` x `n` x `n_years`) for its analysis, which was more concise to implement but so incredibly difficult to debug when there were issues.  If we were to only output a few summary rasters per transition year, we would still have these 3D arrays, and debugging would be much trickier.  Now, there's probably a way to make all this much easier to debug than the prior implementation, but even if we were to do so, we'd still have the unbounded-array-size problem, mentioned in point 1.

I'm also not particularly happy with the sheer number of rasters produced (and the resulting hit to runtime), but I do honestly think that the model will be more useful for our users and easier to debug as a result, and I'm also glad to say that this should never ever run out of memory on consumer hardware.

Reply:
Those sound like really good reasons to do it this way! This could be another decision that's worth recording for posterity, either in a big comment or in the user's guide (could pretty much copy-and-paste what you wrote above)

Reply:
Added some notes about a couple of things that have been discussed about the model relating to its implementation in ca89ccb9, much of which came from the discussion in this thread.